huaweicloud/services/iotda/resource_huaweicloud_iotda_batchtask.go

```diff
@@ -0,0 +1,420 @@
+	req := model.CreateBatchTaskRequest{
+		Body: &model.CreateBatchTask{
+			TaskName:       d.Get("name").(string),
+			TaskType:       d.Get("type").(string),
+			AppId:          utils.StringIgnoreEmpty(d.Get("space_id").(string)),
+			Targets:        utils.ExpandToStringListPointer(d.Get("targets").([]interface{})),
+			TargetsFilter:  utils.RemoveNil(d.Get("targets_filter").(map[string]interface{})),
+			DocumentSource: fileId,
+		},
+	}
+	return &req
+}
+
+func deleteBatchTaskFile(client *iotdav5.IoTDAClient, fileId string) error {
+	deleteOpts := &model.DeleteBatchTaskFileRequest{
+		FileId: fileId,
+	}
+	_, err := client.DeleteBatchTaskFile(deleteOpts)
+	if err != nil {
+		return fmt.Errorf("error deleting IoTDA batch task file after failed creation of IoTDA batch task: %s", err)
+	}
+
+	return nil
+}
+
+func batchTaskStateRefreshFunc(client *iotdav5.IoTDAClient, taskId string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		// There is no need to handle pagination related parameters, as it only applies to the structure of subtasks.
+		getOpts := &model.ShowBatchTaskRequest{
+			TaskId: taskId,
+		}
+		resp, err := client.ShowBatchTask(getOpts)
+		if err != nil {
+			if _, ok := err.(golangsdk.ErrDefault404); ok {
+				return nil, "COMPLETED", err
+			}
+			return nil, "ERROR", err
+		}
+
+		status := *resp.Batchtask.Status
+
+		targetStatus := []string{
+			BatchTaskSuccess, BatchTaskFail, BatchTaskPartialSuccess,
+		}
+		if utils.StrSliceContains(targetStatus, status) {
+			return resp, "COMPLETED", nil
+		}
+
+		pendingStatus := []string{
+			BatchTaskInitializing, BatchTaskWaiting, BatchTaskProcessing,
+		}
+		if utils.StrSliceContains(pendingStatus, status) {
+			return resp, "PENDING", nil
+		}
+
+		return resp, *resp.Batchtask.Status, nil
+	}
+}
+
+func resourceBatchTaskRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	isDerived := withDerivedAuth(cfg, region)
+	client, err := cfg.HcIoTdaV5Client(region, isDerived)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	var (
+		allTaskDetails []model.TaskDetail
+		targetResp     *model.ShowBatchTaskResponse
+		limit          = int32(50)
+		offset         int32
+	)
+
+	// The purpose of pagination is to get all subtasks under the batch task.
+	// The pagination parameter only takes effect on the structure of subtask details(task_details).
+	for {
+		getOpts := &model.ShowBatchTaskRequest{
+			TaskId: d.Id(),
+			Limit:  utils.Int32(limit),
+			Offset: &offset,
+		}
+		resp, getErr := client.ShowBatchTask(getOpts)
+		if getErr != nil {
+			return common.CheckDeletedDiag(d, getErr, "error querying IoTDA batch task")
+		}
+
+		if targetResp == nil {
+			targetResp = resp
+		}
+
+		allTaskDetails = append(allTaskDetails, *resp.TaskDetails...)
+		if len(*resp.TaskDetails) == 0 {
+			break
+		}
+
+		offset += limit
+	}
+
```

Comment:
Add preCheck: whether targetResp is nil, and whether targetResp.Batchtask is nil.

Reply:
fixed.